samples/net/sockets/http_server/src/server.c

```diff
@@ -0,0 +1,889 @@
+			}
+
+			if (!(ctx->fds[i].revents & POLLIN))
+				continue;
+
+			if (i == 0) {
+				int new_socket =
+					accept_new_client(ctx->server_fd);
+
+				for (int j = 2; j < MAX_CLIENTS + 2; j++) {
+					if (ctx->fds[j].fd != 0)
+						continue;
+
+					ctx->fds[j].fd = new_socket;
+					ctx->fds[j].events = POLLIN;
+					ctx->clients[j - 2].client_fd =
+						new_socket;
+					ctx->clients[j - 2].offset = 0;
+					ctx->clients[j - 2].is_http2 = 0;
+					ctx->clients[j - 2]
+						.streams[0]
+						.stream_state = -1;
+
+					if (j > ctx->num_clients + 1)
+						ctx->num_clients++;
+
+					break;
+				}
+				continue;
+			}
+
+			if (i == 1) {
+#ifdef __ZEPHYR__
+				eventfd_read(ctx->event_fd, &value);
+#else
+				read(ctx->event_fd, &value, sizeof(value));
+#endif
+				printf("Received stop event. exiting ..\n");
+
+				return 0;
+			}
+
+			int valread = recv(ctx->clients[i - 2].client_fd,
+				ctx->clients[i - 2].buffer +
+					ctx->clients[i - 2].offset,
+				sizeof(ctx->clients[i - 2].buffer) -
+					ctx->clients[i - 2].offset,
+				0);
+
+			if (valread < 0) {
+				LOG_ERR("ERROR reading from socket");
+				close_client_connection(ctx, i - 2);
+				continue;
+			}
+
+			if (valread == 0) {
+				LOG_INF("Connection closed by peer.\n");
+				close_client_connection(ctx, i - 2);
+				continue;
+			}
+
+			ctx->clients[i - 2].offset += valread;
+			ctx->clients[i - 2].server_state = INITIAL_STATE;
+			handle_http_request(ctx, &ctx->clients[i - 2], i - 2);
+		}
+	} while (ctx->infinite == 1);
+	return 0;
+}
+
+void close_client_connection(
+	struct http2_server_ctx *ctx_server, int client_index)
+{
+	client_index += 2;
+	close(ctx_server->fds[client_index].fd);
+	ctx_server->fds[client_index].fd = 0;
+	ctx_server->fds[client_index].events = 0;
+	ctx_server->fds[client_index].revents = 0;
+
+	if (client_index == ctx_server->num_clients) {
+		while (ctx_server->num_clients > 0 &&
+			ctx_server->fds[ctx_server->num_clients].fd == 0) {
+			--ctx_server->num_clients;
+		}
+	}
+}
+
+int handle_http_request(struct http2_server_ctx *ctx_server,
+	struct http2_client_ctx *ctx_client, int client_index)
+{
+	int ret = -EINVAL;
+
+	do {
+		switch (ctx_client->server_state) {
+		case INITIAL_STATE:
+			ret = handle_initial_state(ctx_client);
+			break;
+		case HTTP_PREFACE:
+			ret = handle_http_preface(ctx_client);
+			break;
+		case HTTP1_REQUEST:
+			ret = handle_http1_request(
+				ctx_server, ctx_client, client_index);
+			break;
+		case HTTP2_FRAME_SETTINGS:
+			ret = handle_http2_frame_setting(ctx_client);
+			break;
+		case HTTP2_FRAME_WINDOW_UPDATE:
+			ret = handle_http2_frame_window_update(ctx_client);
+			break;
+		case STREAMING:
+			ret = handle_http2_streaming(ctx_client);
+			break;
+		case HTTP2_FRAME_GOAWAY:
+			ret = handle_http2_frame_goaway(
+				ctx_server, ctx_client, client_index);
+			break;
+		case CLOSING:
+			break;
+		default:
+			LOG_ERR("Unknown state.\n");
+			close_client_connection(ctx_server, client_index);
+		}
+
```

Comment:
You should break the loop if `ret < 0`. And handle `-EAGAIN` as a special case - it's not an error, but an indication that more data is expected.